File: Session_12_BranchAndBound/multi_language_implementations/knapsack_branch_bound.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type Node struct {
	level  int
	profit int
	weight int
	bound  float64
}

type Item struct {
	id     int
	weight int
	value  int
}

func calculateBound(u Node, n, W int, items []Item) float64 {
	if u.weight >= W {
		return 0
	}

	profit_bound := float64(u.profit)
	j := u.level + 1
	total_weight := u.weight

	for j < n && total_weight+items[j].weight <= W {
		total_weight += items[j].weight
		profit_bound += float64(items[j].value)
		j++
	}

	if j < n {
		profit_bound += float64(W-total_weight) * (float64(items[j].value) / float64(items[j].weight))
	}

	return profit_bound
}

func solveKnapsackBB(W int, items []Item, n int) int {
	// Sort items by ratio descending
	sort.Slice(items, func(i, j int) bool {
		r1 := float64(items[i].value) / float64(items[i].weight)
		r2 := float64(items[j].value) / float64(items[j].weight)
		return r1 > r2
	})

	var q []Node

	var u Node
	u.level = -1
	u.profit = 0
	u.weight = 0
	u.bound = calculateBound(u, n, W, items)

	maxProfit := 0
	q = append(q, u)

	fmt.Println("Branch and Bound Search Steps:")

	for len(q) > 0 {
		u = q[0]
		q = q[1:]

		var v Node
		if u.level == -1 {
			v.level = 0
		} else if u.level == n-1 {
			continue
		} else {
			v.level = u.level + 1
		}

		// --- BRANCH 1: Include next item ---
		v.weight = u.weight + items[v.level].weight
		v.profit = u.profit + items[v.level].value

		if v.weight <= W && v.profit > maxProfit {
			maxProfit = v.profit
			fmt.Printf("  Updated max profit to: %d (Level %d, Weight %d)\n", maxProfit, v.level, v.weight)
		}

		v.bound = calculateBound(v, n, W, items)

		if v.bound > float64(maxProfit) {
			q = append(q, v)
			fmt.Printf("    Branch WITH item %d enqueued. Bound = %.2f\n", items[v.level].id, v.bound)
		} else {
			fmt.Printf("    Branch WITH item %d pruned. Bound (%.2f) <= MaxProfit (%d)\n", items[v.level].id, v.bound, maxProfit)
		}

		// --- BRANCH 2: Exclude next item ---
		v.weight = u.weight
		v.profit = u.profit
		v.bound = calculateBound(v, n, W, items)

		if v.bound > float64(maxProfit) {
			q = append(q, v)
			fmt.Printf("    Branch WITHOUT item %d enqueued. Bound = %.2f\n", items[v.level].id, v.bound)
		} else {
			fmt.Printf("    Branch WITHOUT item %d pruned. Bound (%.2f) <= MaxProfit (%d)\n", items[v.level].id, v.bound, maxProfit)
		}
	}

	return maxProfit
}

func main() {
	W := 10
	items := []Item{
		{1, 2, 40},
		{2, 5, 30},
		{3, 10, 50},
		{4, 5, 10},
	}
	n := len(items)

	fmt.Println("=== Branch and Bound: 0/1 Knapsack ===")
	fmt.Printf("Capacity: %d\n\n", W)

	maxProfit := solveKnapsackBB(W, items, n)

	fmt.Printf("\nMaximum Profit: %d\n", maxProfit)

	fmt.Println()
	fmt.Print("Press Enter to exit...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}

File: Session_12_BranchAndBound/multi_language_implementations/knapsack_branch_bound.rs
use std::collections::VecDeque;
use std::io::{self, BufRead};

#[derive(Clone, Copy, Debug)]
struct Node {
    level: i32,
    profit: i32,
    weight: i32,
    bound: f64,
}

#[derive(Clone, Copy, Debug)]
struct Item {
    id: i32,
    weight: i32,
    value: i32,
}

fn calculate_bound(u: Node, n: usize, w_cap: i32, items: &[Item]) -> f64 {
    if u.weight >= w_cap {
        return 0.0;
    }

    let mut profit_bound = u.profit as f64;
    let mut j = (u.level + 1) as usize;
    let mut total_weight = u.weight;

    while j < n && total_weight + items[j].weight <= w_cap {
        total_weight += items[j].weight;
        profit_bound += items[j].value as f64;
        j += 1;
    }

    if j < n {
        profit_bound += (w_cap - total_weight) as f64 * (items[j].value as f64 / items[j].weight as f64);
    }

    profit_bound
}

fn solve_knapsack_bb(w_cap: i32, items: &mut [Item], n: usize) -> i32 {
    // Sort items by value-to-weight ratio descending
    items.sort_by(|a, b| {
        let r1 = a.value as f64 / a.weight as f64;
        let r2 = b.value as f64 / b.weight as f64;
        r2.partial_cmp(&r1).unwrap()
    });

    let mut q = VecDeque::new();

    let mut u = Node {
        level: -1,
        profit: 0,
        weight: 0,
        bound: 0.0,
    };
    u.bound = calculate_bound(u, n, w_cap, items);

    let mut max_profit = 0;
    q.push_back(u);

    println!("Branch and Bound Search Steps:");

    while let Some(curr_u) = q.pop_front() {
        let mut v = Node {
            level: 0,
            profit: 0,
            weight: 0,
            bound: 0.0,
        };

        if curr_u.level == -1 {
            v.level = 0;
        } else if curr_u.level == (n as i32) - 1 {
            continue; // Leaf node reached
        } else {
            v.level = curr_u.level + 1;
        }

        // --- BRANCH 1: Include next item ---
        let item_idx = v.level as usize;
        v.weight = curr_u.weight + items[item_idx].weight;
        v.profit = curr_u.profit + items[item_idx].value;

        if v.weight <= w_cap && v.profit > max_profit {
            max_profit = v.profit;
            println!("  Updated max profit to: {} (Level {}, Weight {})", max_profit, v.level, v.weight);
        }

        v.bound = calculate_bound(v, n, w_cap, items);

        if v.bound > max_profit as f64 {
            q.push_back(v);
            println!("    Branch WITH item {} enqueued. Bound = {:.2}", items[item_idx].id, v.bound);
        } else {
            println!("    Branch WITH item {} pruned. Bound ({:.2}) <= MaxProfit ({})", items[item_idx].id, v.bound, max_profit);
        }

        // --- BRANCH 2: Exclude next item ---
        let mut v_exclude = Node {
            level: v.level,
            profit: curr_u.profit,
            weight: curr_u.weight,
            bound: 0.0,
        };
        v_exclude.bound = calculate_bound(v_exclude, n, w_cap, items);

        if v_exclude.bound > max_profit as f64 {
            q.push_back(v_exclude);
            println!("    Branch WITHOUT item {} enqueued. Bound = {:.2}", items[item_idx].id, v_exclude.bound);
        } else {
            println!("    Branch WITHOUT item {} pruned. Bound ({:.2}) <= MaxProfit ({})", items[item_idx].id, v_exclude.bound, max_profit);
        }
    }

    max_profit
}

fn main() {
    let w_cap = 10;
    let mut items = [
        Item { id: 1, weight: 2, value: 40 },
        Item { id: 2, weight: 5, value: 30 },
        Item { id: 3, weight: 10, value: 50 },
        Item { id: 4, weight: 5, value: 10 },
    ];
    let n = items.len();

    println!("=== Branch and Bound: 0/1 Knapsack ===");
    println!("Capacity: {}\n", w_cap);

    let max_profit = solve_knapsack_bb(w_cap, &mut items, n);

    println!("\nMaximum Profit: {}", max_profit);

    println!();
    println!("Press Enter to exit...");
    let mut iterator = io::stdin().lock().lines();
    let _ = iterator.next();
}
